### PR TITLE
Detect premature worker load error

### PR DESCRIPTION
**Pull request by Rob--W**
Rejects the ready promise when the worker fails to load, e.g. due to network errors.

### CODE REVIEW

**Comment by kyoshino** [comment]
This PR is required for Firefox 45+, because `new Worker()` no longer throws a `SecurityError` when the worker failed to load due to an origin issue, according to [Bug 1260388](https://bugzilla.mozilla.org/show_bug.cgi?id=1260388) and [Bug 1260961](https://bugzilla.mozilla.org/show_bug.cgi?id=1260961). Any load errors must be caught with the `error` handler, otherwise the fallback function will never get called.

File: src/display/api.js [hunk 1]
```diff
@@ -1302,11 +1302,25 @@ var PDFWorker = (function PDFWorkerClosure() {
           // https://bugzilla.mozilla.org/show_bug.cgi?id=683280
           var worker = new Worker(workerSrc);
           var messageHandler = new MessageHandler('main', 'worker', worker);
+          var terminateEarly = function() {
+            this._readyCapability.reject(new Error('Worker was destroyed'));
+            messageHandler.destroy();
+            worker.removeEventListener('error', onWorkerError);
+            worker.terminate();
```
Comment (kyoshino) [inline src/display/api.js]: `this._setupFakeWorker();` is needed here to call the fallback function.

File: src/display/api.js [hunk 2]
```diff
@@ -1302,11 +1302,25 @@ var PDFWorker = (function PDFWorkerClosure() {
           // https://bugzilla.mozilla.org/show_bug.cgi?id=683280
           var worker = new Worker(workerSrc);
           var messageHandler = new MessageHandler('main', 'worker', worker);
+          var terminateEarly = function() {
+            this._readyCapability.reject(new Error('Worker was destroyed'));
```
Comment (kyoshino) [inline src/display/api.js]: `this._setupFakeWorker();` is needed here to call the fallback function.

**Comment by Snuffleupagus** [comment]
> This PR is required for Firefox 45+, because new Worker() no longer throws a SecurityError when the worker failed to load due to an origin issue, according to Bug 1260388 and Bug 1260961.

If I'm understanding the bugs correctly, this issue *only* affect custom deployments of PDF.js, and *not* the version built-in to Firefox!?
That seems like an important distinction to me, since it means we won't need to uplift the fix to Firefox.

**Comment by kyoshino** [comment]
I think the build-in version in Firefox is NOT affected.

File: src/display/api.js (hunk 2 above)
Comment (Rob--W) [inline src/display/api.js]: Done; I've overwritten the previous commit, with the following changes (see PR for full change set).

```diff
@@ -1303,15 +1303,22 @@ var PDFWorker = (function PDFWorkerClosure() {
           var worker = new Worker(workerSrc);
           var messageHandler = new MessageHandler('main', 'worker', worker);
           var terminateEarly = function() {
-            this._readyCapability.reject(new Error('Worker was destroyed'));
-            messageHandler.destroy();
             worker.removeEventListener('error', onWorkerError);
+            messageHandler.destroy();
             worker.terminate();
+            if (this.destroyed) {
+              this._readyCapability.reject(new Error('Worker was destroyed'));
+            } else {
+              // Fall back to fake worker if the termination is caused by an
+              // error (e.g. NetworkError / SecurityError).
+              this._setupFakeWorker();
+            }
           }.bind(this);
 
           var onWorkerError = function(event) {
-            if (!event.message && !this.destroyed && !this._webWorker) {
-              // Worker failed to initialize due to a load error.
+            if (!this._webWorker) {
+              // Worker failed to initialize due to an error. Clean up and fall
+              // back to the fake worker.
               terminateEarly();
             }   
           }.bind(this);
```

**Comment by yurydelendik** [comment]
Looks good.

/botio test

**Comment by yurydelendik** [comment]
Thank you for the patch.

**Comment by kyoshino** [comment]
Awesome, thank you!

**Comment by kyoshino** [comment]
@yurydelendik Any ETA on PDF.js v1.4.187?

**Comment by yurydelendik** [comment]
> Any ETA on PDF.js v1.4.187?

@kyoshino you can download library from the mozilla/pdfjs-dist repo. the viewer package will be available as soon we are release next beta. You can build it yourself using `gulp generic`.

**Comment by kyoshino** [comment]
Oh I see, thanks. I'll mention it in my site compat doc.

**Pull request merged**